File: Limbus/limbus_core/backend.fs
module backend
open common


[<AbstractClass>]
type Backend(iCode : iCode, symTab : SymTab) = 
    let mutable iCode = iCode
    let mutable symTab = symTab

    member this.message_handler = MessegeHandller()    
    
    interface  MessageProducer with 
        member this.add_message_listener(ml: MessageListener) : unit = 
            this.message_handler.add_message_listener(ml)
        member this.remove_message_listener(ml : MessageListener) : unit = 
            this.message_handler.remove_message_listener(ml)
        member this.send_message(m : Message) : unit = 
            this.message_handler.send_message(m)
            
    abstract member process : iCode -> SymTab -> unit
    

[<Class>]
type Interpreter(iCode : iCode, symTab : SymTab) = 
    let mutable iCode = iCode
    let mutable symTab = symTab
    
    member this.message_handler = MessegeHandller()    
    member val execution_count = 0 with set, get
    member val runtime_errres = 0 with set, get
    
    
    interface  MessageProducer with 
         member this.add_message_listener(ml: MessageListener) : unit = 
             this.message_handler.add_message_listener(ml)
         member this.remove_message_listener(ml : MessageListener) : unit = 
             this.message_handler.remove_message_listener(ml)
         member this.send_message(m : Message) : unit = 
             this.message_handler.send_message(m)
         
    member this.send_message(m : Message) : unit = 
        this.message_handler.send_message(m)
         
         
    member this.lprocess(iCode : iCode, symTab : SymTab) : unit =
        this.execution_count <- 0
        this.runtime_errres <- 0
        //sendmessage
        let es = string(this.execution_count)
        let rs = string(this.runtime_errres)
        
        let m = {msg_type = INTERPRETER_SUMMARY ; body = MSGS [es ; rs];}
        this.send_message(m)
               
                     

File: Limbus/limbus_core/backend/backend.py
# -*- coding: utf-8 -*-
from abc import ABCMeta, abstractmethod
import sys
sys.path.append("limbus_core")

from message import MessageProducer, MessageHandler


class Backend(MessageProducer, metaclass=ABCMeta):
    def __init__(self):
        self.iCode = None
        self.symtab = None
        self.message_handler = MessageHandler()

    @abstractmethod
    def process(self, icode, symtab):
        raise NotImplementedError()

    # delegate
    def add_message_listener(self, listener):
        self.message_handler.add_message_listener(listener)

    def remove_message_listener(self, listener):
        self.message_handler.remove_message_listener(listener)

    def send_message(self, message):
        self.message_handler.send_message(message)

class CodeGeneretor(Backend):
    def __init__(self):
        self.execution_count = 0
        self.runtime_error = 0
        super().__init__()

    def process(self, icode, symtab):
        ec = 0
        msg = Message(MessageType.COMPILER_SUMMARY, ec)
        self.send_message(msg)

